var/spack/repos/builtin/packages/py-torch/package.py

```diff
@@ -597,6 +597,13 @@ def enable_or_disable(variant, keyword="USE", var=None, newer=False):
             # https://github.com/pytorch/pytorch/issues/60332
             # env.set("USE_SYSTEM_XNNPACK", "ON")
 
+    def setup_run_environment(self, env):
+        # The cmake_prefix_paths property from below doesn't end up in the run
+        # environment that one gets with spack load py-torch
+        env.prepend_path("CMAKE_PREFIX_PATH", self.cmake_prefix_paths[0])
```

Comment:
I assume this is so you can build packages outside of Spack that use PyTorch and CMake? 

From examining the `py-torch` package, I can say that the base prefix is already part of `modules:prefix_inspections` configuration and should be added by default if you e.g. `spack load py-torch`. 

Given

> This one also has the problem that by populating CMAKE_PREFIX_PATH in setup_run_environment there is still the one that simply sets the prefix

I think you mean that some dependency of py-torch is overriding CMAKE_PREFIX_PATH by setting it rather than appending/prepending: I think it would be best to find that dependency and edit it. I assume it would be easy to find that dependency, because if you remove this line and `spack load py-torch`, it should appear in CMAKE_PREFIX_PATH.

Side note: For adding any of the other prefixes to CMAKE_PREFIX_PATH, I'll note that the configuration is a bit more-involved if you want to set other values (https://spack.readthedocs.io/en/latest/module_file_support.html#selection-by-anonymous-specs might be helpful for that).

Reply:
> I think you mean that some dependency of py-torch is overriding CMAKE_PREFIX_PATH by setting it rather than appending/prepending:

I am not sure whether it is one of the dependencies, or whether it is the `prefix_inspections` configuration. What I was referring to here is the fact that something sets `CMAKE_PREFIX_PATH` to the base prefix of `py-torch` with `spack load py-torch`. That is not the correct place in this case. If I now use this patch, I get a second path added to `CMAKE_PREFIX_PATH` for `py-torch` (pointing to the correct place).


Is the `modules:prefix_inspections` you are reffering to the `etc/spack/defaults/modules.yaml`? Or is there another one with more specific overrides?

Is there an easier way to debug this further than just stepping through `spack load py-toarch` with pdb?

Reply:
Thanks for the clarification. [4] would be very useful I think. I am not entirely sure how easy it is to implement in the end, since the expected behavior might differ slightly from "normal spack" behavior. E.g.
- Is `spack load --context=build` effectively the same as `spack load --only=dependencies` + `setup_dependent_build_environment`. Which would allow you to build against a package, but not necessarily run in the same environement.
- Or is it effectively the same as `spack load` (i.e. runtime environment) with `setup_dependent_build_environment`. This would basically allow to build and run, but is probably mixing quite a few concerns for spack.

(Taking `setup_dependent_build_environment` as a placeholder that I know of, it might not be the correct/actual thing that needs to happen.)